module/zfs/dsl_dataset.c

```diff
@@ -727,26 +802,41 @@ dsl_dataset_disown(dsl_dataset_t *ds, void *tag)
 	ASSERT(ds->ds_dbuf != NULL);
 
 	mutex_enter(&ds->ds_lock);
+	if (ds->ds_dir && ds->ds_dir->dd_crypto_obj) {
+		(void) spa_keystore_remove_mapping(ds->ds_dir->dd_pool->dp_spa,
+		    ds, ds);
```

Comment:
As part of the `dsl_dataset_hold_crypt()` change I mentioned above, the `dsl_dataset_rele()` and `dsl_dataset_disown()` will also accept flags to properly manage releasing the keys.